cmd/operator-opamp-bridge/metrics/reporter.go

```diff
@@ -0,0 +1,174 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/shirou/gopsutil/process"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+	"go.opentelemetry.io/otel/metric/instrument/asyncint64"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	otelresource "go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+)
+
+// MetricReporter is a metric reporter that collects Agent metrics and sends them to an
+// OTLP/HTTP destination.
+type MetricReporter struct {
+	logger types.Logger
+
+	meter           metric.Meter
+	meterShutdowner func()
+	done            chan struct{}
+
+	// The Agent's process.
+	process *process.Process
+
+	// Some example metrics to report.
+	processMemoryPhysical asyncint64.Gauge
+	processCpuTime        asyncfloat64.Counter
+}
+
+// NewMetricReporter creates an OTLP gRPC client to the destination address supplied by the server.
+// TODO: do more validation on the endpoint, allow for http.
+// TODO: set global provider and add more metrics to be reported.
+func NewMetricReporter(
+	logger types.Logger,
+	dest *protobufs.TelemetryConnectionSettings,
+	agentType string,
+	agentVersion string,
+	instanceId ulid.ULID,
+) (*MetricReporter, error) {
+
+	if dest.DestinationEndpoint == "" {
+		err := fmt.Errorf("metric destination must specify DestinationEndpoint")
+		return nil, err
+	}
+
+	// Create OTLP/grpc metric exporter.
```

Comment:
OpAMP spec says that own metrics should be reported using OTLP/HTTP: https://github.com/open-telemetry/opamp-spec/blob/main/specification.md#own-telemetry-reporting

Reply:
is there a reason it specifies HTTP over grpc? I tend to always use gRPC when it's available.

Reply:
2 reasons:
1. OpAMP itself has no dependency on gRPC, so it was seen as an unnecessary additional burden for agents that don't use gRCP for any other purpose.
2. OTLP/HTTP is slightly more performant than OTLP/gRPC (particularly for small payloads). Generally, I am not aware of any use case where OTLP/gRPC is preferable over OTLP/HTTP.